File: Arrays/0033-search-in-rotated-sorted-array.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l, r = 0, len(nums) - 1

        while l <= r:
            mid = (l + r) // 2

            if nums[mid] == target:
                return mid

            # Check which side is sorted
            if nums[l] <= nums[mid]:
                # Left side sorted hai
                if nums[l] <= target < nums[mid]:
                    r = mid - 1 # target left me hai
                else:
                    l = mid + 1 # target right me hai
            else:
                # Right side sorted hai
                if nums[mid] < target <= nums[r]:
                    l = mid + 1 # target right me hai
                else:
                    r = mid - 1 # target left me hai

        return -1

File: Arrays/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0, right = nums.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] == target) return mid;

            // Check if left half is sorted
            if (nums[left] <= nums[mid]) {
                if (nums[left] <= target && target < nums[mid]) {
                    right = mid - 1; // Target lies in the sorted left half
                } else {
                    left = mid + 1;  // Search in the right half
                }
            } 
            // Right half is sorted
            else {
                if (nums[mid] < target && target <= nums[right]) {
                    left = mid + 1;  // Target lies in the sorted right half
                } else {
                    right = mid - 1; // Search in the left half
                }
            }
        }

        return -1;
    }
};
